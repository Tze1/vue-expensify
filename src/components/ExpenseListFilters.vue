<template>
  <div class="expenselistfilters component">
    <h4 class="expenselistfilters-title">Filters</h4>
    <div class="expenselistfilters-controls">
      <input
        type="text"
        name="text"
        placeholder="filter by text"
        v-model="filters.text"
        @input="onTextInput"
      />
      <Datepicker
        name="startDate"
        format="MM/dd/yyyy"
        :value="startDateValue"
        @selected="onStartDateChange"
        @cleared="onStartDateChange"
        input-class="datepciker-input"
        :clear-button="true"
      />
      <Datepicker
        name="endDate"
        format="MM/dd/yyyy"
        :value="endDateValue"
        @selected="onEndDateChange"
        @cleared="onEndDateChange"
        input-class="datepciker-input"
        :clear-button="true"
      />
      <select
        name="sortBy"
        v-model="filters.sortBy"
        @change="onSortByChange"
      >
        <option value="date">Sort by Date</option>
        <option value="amount">Sort by Amount</option>
      </select>
    </div>
  </div>
</template>

<script>
  import moment from 'moment';
  import Datepicker from 'vuejs-datepicker';

  export default {
    name: 'ExpenseListFilters',
    props: {
      filters: {
        type: Object,
        required: true,
      },
    },
    computed: {
      startDateValue: function () {
        return this.filters.startDate ?
          moment(this.filters.startDate).toDate() :
          undefined;
      },
      endDateValue: function () {
        return this.filters.endDate ?
          moment(this.filters.endDate).toDate() :
          undefined;
      }
    },
    methods: {
      onTextInput (e) {
        console.log('[ExpenseListFilters onTextInput event:', e);
        this.$store.dispatch('setFilterText', e.target.value);
      },
      onStartDateChange (e) {
        if (e) {
          this.$store.dispatch('setFilterStartDate', moment(e.getTime()));
        } else {
          this.$store.dispatch('setFilterStartDate', undefined);
        }
      },
      onEndDateChange (e) {
        if (e) {
          this.$store.dispatch('setFilterEndDate', moment(e.getTime()));
        } else {
          this.$store.dispatch('setFilterEndDate', undefined);
        }
      },
      onSortByChange (e) {
        this.$store.dispatch('setFilterSortBy', e.target.value);
      }
    },
    components: {
      Datepicker
    }
  };
</script>

<style scoped lang="scss">
  @import '../styles/app.scss';
  .expenselistfilters.component {
      font-size: 1.4rem;

    .expenselistfilters-title {
      margin-bottom: $space-xs;
      margin-top: $space-xs;
    }

    .expenselistfilters-controls {
      display: flex;
      flex-flow: column;
      font-size: 1.4rem;

      > * {
        margin-bottom: $space-xs;

        *:last-child {
          margin-bottom: 0;
        }
      }

      > input,
      > select {
        border: 0;
        font-size: 1.4rem;
        overflow: hidden;
        padding: $space-xs;

        > option {
          font-size: 1.4rem;
        }
      }

      @include mq("tablet") {
        flex-flow: row nowrap;
        justify-content: space-between;

        > input,
        > select {
          margin-bottom: 0;
        }
      }
    }
  }
</style>
