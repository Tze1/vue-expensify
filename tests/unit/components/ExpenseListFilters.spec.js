import { mount } from '@vue/test-utils';
import ExpenseListFilters from '@/components/ExpenseListFilters';
import moment from 'moment';
import Datepicker from 'vuejs-datepicker';
import { defaultFilters, altFilters } from '../../fixtures/testFilters';

describe('ExpenseListFilters.vue', () => {
  const onTextInputStub = jest.fn();
  const onStartDateChangeStub = jest.fn();
  const onEndDateChangeStub = jest.fn();
  const onSortByChangeStub = jest.fn();
  const wrapper = mount(ExpenseListFilters, {
    propsData: {
      filters: defaultFilters,
    },
  });
  wrapper.setMethods({
    onTextInput: onTextInputStub,
    onStartDateChange: onStartDateChangeStub,
    onEndDateChange: onEndDateChangeStub,
    onSortByChange: onSortByChangeStub,
  });

  it('renders ExpenseListFilters component properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders ExpenseListFilters component properly with altered filters', () => {
    wrapper.setProps({ filters: altFilters });

    expect(wrapper).toMatchSnapshot();
  });

  it('calls onTextInput correctly on text input', () => {
    const testText = 'a';
    const textInput = wrapper.find('input[name="text"]');

    textInput.setValue(testText);

    // expect(onTextInputStub).toHaveBeenCalled();
    expect(onTextInputStub).toHaveBeenCalledWith(
      expect.objectContaining({
        target: expect.objectContaining({ value: testText })
      })
    );
  });

  it('calls onStartDateChange correctly on datepicker selection', () => {
    const testDate = new Date('2018-10-01');
    const startDatepicker = wrapper.find(Datepicker);

    startDatepicker.vm.$emit('selected', testDate);

    expect(onStartDateChangeStub).toHaveBeenCalledWith(testDate);
  });

  it('calls onStartDateChange correctly on datepicker clear', () => {
    const startDatepicker = wrapper.find(Datepicker);

    startDatepicker.vm.$emit('cleared', null);

    expect(onStartDateChangeStub).toHaveBeenCalledWith(null);
  });

  it('calls onEndDateChange correctly on datepicker selection', () => {
    const testDate = new Date('2018-11-01');
    const endDatepicker = wrapper.findAll(Datepicker).at(1);

    endDatepicker.vm.$emit('selected', testDate);

    expect(onEndDateChangeStub).toHaveBeenCalledWith(testDate);
  });

  it('calls onEndDateChange correctly on datepicker clear', () => {
    const endDatepicker = wrapper.findAll(Datepicker).at(1);

    endDatepicker.vm.$emit('cleared', null);

    expect(onEndDateChangeStub).toHaveBeenCalledWith(null);
  });

  it('calls onSortByChange correctly on sort-by select change', () => {
    const testSortBy = 'amount';
    const sortSelect = wrapper.find('select[name="sortBy"]');

    sortSelect.setValue(testSortBy);

    expect(onSortByChangeStub).toHaveBeenCalledWith(
      expect.objectContaining({
        target: expect.objectContaining({ value: testSortBy })
      })
    );
  });
});
