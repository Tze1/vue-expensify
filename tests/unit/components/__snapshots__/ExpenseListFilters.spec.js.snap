// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseListFilters.vue renders ExpenseListFilters component properly 1`] = `
<div class="expenselistfilters component">
  <h4 class="expenselistfilters-title">Filters</h4>
  <div class="expenselistfilters-controls"><input type="text" name="text" placeholder="filter by text">
    <div class="vdp-datepicker">
      <div class="">
        <!----> <input type="text" name="startDate" clear-button="true" readonly="readonly" autocomplete="off" class="datepciker-input"> <span class="vdp-datepicker__clear-button"><span class=""><i><span>×</span></i></span></span> </div>
      <div class="vdp-datepicker__calendar" style="display: none;">
        <header><span class="prev">&lt;</span> <span class="day__month_btn up">Sep 2018</span> <span class="next">&gt;</span></header>
        <div class=""><span class="cell day-header">Sun</span><span class="cell day-header">Mon</span><span class="cell day-header">Tue</span><span class="cell day-header">Wed</span><span class="cell day-header">Thu</span><span class="cell day-header">Fri</span><span class="cell day-header">Sat</span> <span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day weekend sat">1</span><span class="cell day weekend sun">2</span><span class="cell day">3</span><span class="cell day">4</span><span class="cell day">5</span><span class="cell day">6</span><span class="cell day">7</span><span class="cell day weekend sat">8</span><span class="cell day weekend sun">9</span><span class="cell day">10</span><span class="cell day selected">11</span><span class="cell day">12</span><span class="cell day">13</span><span class="cell day">14</span><span class="cell day weekend sat">15</span><span class="cell day weekend sun">16</span><span class="cell day">17</span><span class="cell day">18</span><span class="cell day">19</span><span class="cell day">20</span><span class="cell day">21</span><span class="cell day weekend sat">22</span><span class="cell day weekend sun">23</span><span class="cell day">24</span><span class="cell day">25</span><span class="cell day">26</span><span class="cell day">27</span><span class="cell day">28</span><span class="cell day weekend sat">29</span><span class="cell day weekend sun">30</span></div>
      </div>
      <div class="vdp-datepicker__calendar" style="display: none;">
        <header><span class="prev">&lt;</span> <span class="month__year_btn up">2018</span> <span class="next">&gt;</span></header> <span class="cell month">January</span><span class="cell month">February</span><span class="cell month">March</span><span class="cell month">April</span><span class="cell month">May</span><span class="cell month">June</span><span class="cell month">July</span><span class="cell month">August</span><span class="cell month selected">September</span><span class="cell month">October</span><span class="cell month">November</span><span class="cell month">December</span>
      </div>
      <div class="vdp-datepicker__calendar" style="display: none;">
        <header><span class="prev">&lt;</span> <span>2010 - 2019</span> <span class="next">&gt;</span></header> <span class="cell year">2010</span><span class="cell year">2011</span><span class="cell year">2012</span><span class="cell year">2013</span><span class="cell year">2014</span><span class="cell year">2015</span><span class="cell year">2016</span><span class="cell year">2017</span><span class="cell year selected">2018</span><span class="cell year">2019</span>
      </div>
    </div>
    <div class="vdp-datepicker">
      <div class="">
        <!----> <input type="text" name="endDate" clear-button="true" readonly="readonly" autocomplete="off" class="datepciker-input">
        <!---->
      </div>
      <div class="vdp-datepicker__calendar" style="display: none;">
        <header><span class="prev">&lt;</span> <span class="day__month_btn up">Nov 2018</span> <span class="next">&gt;</span></header>
        <div class=""><span class="cell day-header">Sun</span><span class="cell day-header">Mon</span><span class="cell day-header">Tue</span><span class="cell day-header">Wed</span><span class="cell day-header">Thu</span><span class="cell day-header">Fri</span><span class="cell day-header">Sat</span> <span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day">1</span><span class="cell day">2</span><span class="cell day weekend sat">3</span><span class="cell day weekend sun">4</span><span class="cell day">5</span><span class="cell day">6</span><span class="cell day">7</span><span class="cell day">8</span><span class="cell day">9</span><span class="cell day weekend sat">10</span><span class="cell day today weekend sun">11</span><span class="cell day">12</span><span class="cell day">13</span><span class="cell day">14</span><span class="cell day">15</span><span class="cell day">16</span><span class="cell day weekend sat">17</span><span class="cell day weekend sun">18</span><span class="cell day">19</span><span class="cell day">20</span><span class="cell day">21</span><span class="cell day">22</span><span class="cell day">23</span><span class="cell day weekend sat">24</span><span class="cell day weekend sun">25</span><span class="cell day">26</span><span class="cell day">27</span><span class="cell day">28</span><span class="cell day">29</span><span class="cell day">30</span></div>
      </div>
      <div class="vdp-datepicker__calendar" style="display: none;">
        <header><span class="prev">&lt;</span> <span class="month__year_btn up">2018</span> <span class="next">&gt;</span></header> <span class="cell month">January</span><span class="cell month">February</span><span class="cell month">March</span><span class="cell month">April</span><span class="cell month">May</span><span class="cell month">June</span><span class="cell month">July</span><span class="cell month">August</span><span class="cell month">September</span><span class="cell month">October</span><span class="cell month">November</span><span class="cell month">December</span>
      </div>
      <div class="vdp-datepicker__calendar" style="display: none;">
        <header><span class="prev">&lt;</span> <span>2010 - 2019</span> <span class="next">&gt;</span></header> <span class="cell year">2010</span><span class="cell year">2011</span><span class="cell year">2012</span><span class="cell year">2013</span><span class="cell year">2014</span><span class="cell year">2015</span><span class="cell year">2016</span><span class="cell year">2017</span><span class="cell year">2018</span><span class="cell year">2019</span>
      </div>
    </div> <select name="sortBy">
      <option value="date">Sort by Date</option>
      <option value="amount">Sort by Amount</option>
    </select>
  </div>
</div>
`;

exports[`ExpenseListFilters.vue renders ExpenseListFilters component properly with altered filters 1`] = `
<div class="expenselistfilters component">
  <h4 class="expenselistfilters-title">Filters</h4>
  <div class="expenselistfilters-controls"><input type="text" name="text" placeholder="filter by text">
    <div class="vdp-datepicker">
      <div class="">
        <!----> <input type="text" name="startDate" clear-button="true" readonly="readonly" autocomplete="off" class="datepciker-input"> <span class="vdp-datepicker__clear-button"><span class=""><i><span>×</span></i></span></span> </div>
      <div class="vdp-datepicker__calendar" style="display: none;">
        <header><span class="prev">&lt;</span> <span class="day__month_btn up">Dec 1969</span> <span class="next">&gt;</span></header>
        <div class=""><span class="cell day-header">Sun</span><span class="cell day-header">Mon</span><span class="cell day-header">Tue</span><span class="cell day-header">Wed</span><span class="cell day-header">Thu</span><span class="cell day-header">Fri</span><span class="cell day-header">Sat</span> <span class="cell day blank"></span><span class="cell day">1</span><span class="cell day">2</span><span class="cell day">3</span><span class="cell day">4</span><span class="cell day">5</span><span class="cell day weekend sat">6</span><span class="cell day weekend sun">7</span><span class="cell day">8</span><span class="cell day">9</span><span class="cell day">10</span><span class="cell day">11</span><span class="cell day">12</span><span class="cell day weekend sat">13</span><span class="cell day weekend sun">14</span><span class="cell day">15</span><span class="cell day">16</span><span class="cell day">17</span><span class="cell day">18</span><span class="cell day">19</span><span class="cell day weekend sat">20</span><span class="cell day weekend sun">21</span><span class="cell day">22</span><span class="cell day">23</span><span class="cell day">24</span><span class="cell day">25</span><span class="cell day">26</span><span class="cell day weekend sat">27</span><span class="cell day weekend sun">28</span><span class="cell day">29</span><span class="cell day">30</span><span class="cell day selected">31</span></div>
      </div>
      <div class="vdp-datepicker__calendar" style="display: none;">
        <header><span class="prev">&lt;</span> <span class="month__year_btn up">1969</span> <span class="next">&gt;</span></header> <span class="cell month">January</span><span class="cell month">February</span><span class="cell month">March</span><span class="cell month">April</span><span class="cell month">May</span><span class="cell month">June</span><span class="cell month">July</span><span class="cell month">August</span><span class="cell month">September</span><span class="cell month">October</span><span class="cell month">November</span><span class="cell month selected">December</span>
      </div>
      <div class="vdp-datepicker__calendar" style="display: none;">
        <header><span class="prev">&lt;</span> <span>1960 - 1969</span> <span class="next">&gt;</span></header> <span class="cell year">1960</span><span class="cell year">1961</span><span class="cell year">1962</span><span class="cell year">1963</span><span class="cell year">1964</span><span class="cell year">1965</span><span class="cell year">1966</span><span class="cell year">1967</span><span class="cell year">1968</span><span class="cell year selected">1969</span>
      </div>
    </div>
    <div class="vdp-datepicker">
      <div class="">
        <!----> <input type="text" name="endDate" clear-button="true" readonly="readonly" autocomplete="off" class="datepciker-input"> <span class="vdp-datepicker__clear-button"><span class=""><i><span>×</span></i></span></span> </div>
      <div class="vdp-datepicker__calendar" style="display: none;">
        <header><span class="prev">&lt;</span> <span class="day__month_btn up">Jan 1970</span> <span class="next">&gt;</span></header>
        <div class=""><span class="cell day-header">Sun</span><span class="cell day-header">Mon</span><span class="cell day-header">Tue</span><span class="cell day-header">Wed</span><span class="cell day-header">Thu</span><span class="cell day-header">Fri</span><span class="cell day-header">Sat</span> <span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day blank"></span><span class="cell day">1</span><span class="cell day">2</span><span class="cell day selected weekend sat">3</span><span class="cell day weekend sun">4</span><span class="cell day">5</span><span class="cell day">6</span><span class="cell day">7</span><span class="cell day">8</span><span class="cell day">9</span><span class="cell day weekend sat">10</span><span class="cell day weekend sun">11</span><span class="cell day">12</span><span class="cell day">13</span><span class="cell day">14</span><span class="cell day">15</span><span class="cell day">16</span><span class="cell day weekend sat">17</span><span class="cell day weekend sun">18</span><span class="cell day">19</span><span class="cell day">20</span><span class="cell day">21</span><span class="cell day">22</span><span class="cell day">23</span><span class="cell day weekend sat">24</span><span class="cell day weekend sun">25</span><span class="cell day">26</span><span class="cell day">27</span><span class="cell day">28</span><span class="cell day">29</span><span class="cell day">30</span><span class="cell day weekend sat">31</span></div>
      </div>
      <div class="vdp-datepicker__calendar" style="display: none;">
        <header><span class="prev">&lt;</span> <span class="month__year_btn up">1970</span> <span class="next">&gt;</span></header> <span class="cell month selected">January</span><span class="cell month">February</span><span class="cell month">March</span><span class="cell month">April</span><span class="cell month">May</span><span class="cell month">June</span><span class="cell month">July</span><span class="cell month">August</span><span class="cell month">September</span><span class="cell month">October</span><span class="cell month">November</span><span class="cell month">December</span>
      </div>
      <div class="vdp-datepicker__calendar" style="display: none;">
        <header><span class="prev">&lt;</span> <span>1970 - 1979</span> <span class="next">&gt;</span></header> <span class="cell year selected">1970</span><span class="cell year">1971</span><span class="cell year">1972</span><span class="cell year">1973</span><span class="cell year">1974</span><span class="cell year">1975</span><span class="cell year">1976</span><span class="cell year">1977</span><span class="cell year">1978</span><span class="cell year">1979</span>
      </div>
    </div> <select name="sortBy">
      <option value="date">Sort by Date</option>
      <option value="amount">Sort by Amount</option>
    </select>
  </div>
</div>
`;
